distributed/comm/ucx.py

```diff
@@ -100,6 +138,72 @@ def device_array(n):
                     "In order to send/recv CUDA arrays, Numba or RMM is required"
                 )
 
+    # Find the functions `device_concat` and `device_split`
+    try:
+        import cupy
+
+        def device_concat(arys):
+            with cupy.cuda.using_allocator(rmm.rmm_cupy_allocator):
+                arys = [cupy.asarray(e).view("u1") for e in arys]
+                result = cupy.concatenate(arys, axis=None)
+                result_buffer = result.data.mem._owner
+                return result_buffer
+
+        def device_split(ary, indices):
+            with cupy.cuda.using_allocator(rmm.rmm_cupy_allocator):
+                ary = cupy.asarray(ary).view("u1")
+                ary_split = cupy.split(ary, indices)
+                results = []
+                result_buffers = []
+                for e in ary_split:
+                    e2 = cupy.empty_like(e)
+                    cupy.copyto(e2, e)
```

Comment:
Originally was calling `e.copy()` however this calls [`cudaMemcpy`]( https://github.com/cupy/cupy/blob/1ed835dbd735c2756db9a66edfc23a1039ed324e/cupy/cuda/memory.pyx#L336-L337 ), which is synchronous under-the-hood. So have switched to allocating arrays and calling `copyto`, which uses [`cudaMemcpyAsync`]( https://github.com/cupy/cupy/blob/1ed835dbd735c2756db9a66edfc23a1039ed324e/cupy/cuda/memory.pyx#L356-L357 ). Should cutdown on the overhead of copying data into smaller buffers during unpacking.